dhis-2/dhis-services/dhis-service-dxf2/src/test/java/org/hisp/dhis/dxf2/EnrollmentServiceTest.java

```diff
@@ -0,0 +1,248 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.dxf2;
+
+import static org.junit.jupiter.api.Assertions.assertAll;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyBoolean;
+import static org.mockito.Mockito.atLeastOnce;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import java.util.Set;
+
+import org.hisp.dhis.common.CodeGenerator;
+import org.hisp.dhis.dxf2.events.enrollment.AbstractEnrollmentService;
+import org.hisp.dhis.dxf2.events.enrollment.Enrollment;
+import org.hisp.dhis.dxf2.events.event.Event;
+import org.hisp.dhis.dxf2.events.event.EventService;
+import org.hisp.dhis.dxf2.events.params.EnrollmentParams;
+import org.hisp.dhis.dxf2.events.params.TrackedEntityInstanceParams;
+import org.hisp.dhis.dxf2.events.relationship.RelationshipService;
+import org.hisp.dhis.program.Program;
+import org.hisp.dhis.program.ProgramInstance;
+import org.hisp.dhis.program.ProgramStageInstance;
+import org.hisp.dhis.program.ProgramStatus;
+import org.hisp.dhis.relationship.Relationship;
+import org.hisp.dhis.relationship.RelationshipItem;
+import org.hisp.dhis.relationship.RelationshipType;
+import org.hisp.dhis.trackedentity.TrackedEntityAttribute;
+import org.hisp.dhis.trackedentity.TrackedEntityAttributeService;
+import org.hisp.dhis.trackedentity.TrackedEntityInstance;
+import org.hisp.dhis.trackedentity.TrackedEntityType;
+import org.hisp.dhis.trackedentity.TrackerAccessManager;
+import org.hisp.dhis.trackedentityattributevalue.TrackedEntityAttributeValue;
+import org.hisp.dhis.user.CurrentUserService;
+import org.hisp.dhis.user.User;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.Spy;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.springframework.test.util.ReflectionTestUtils;
+
+@ExtendWith( MockitoExtension.class )
+class EnrollmentServiceTest
+{
+    @Mock
+    CurrentUserService currentUserService;
+
+    @Mock
+    EventService eventService;
+
+    @Mock
+    RelationshipService relationshipService;
+
+    @Mock
+    TrackedEntityAttributeService trackedEntityAttributeService;
+
+    @Mock
+    TrackerAccessManager trackerAccessManager;
+
+    @Mock
+    ProgramInstance programInstance;
+
+    @Mock
+    TrackedEntityInstance trackedEntityInstance;
+
+    @Mock
+    User user;
+
+    @Spy
+    AbstractEnrollmentService enrollmentService;
+
+    String enrollmentId = CodeGenerator.generateUid();
+
+    @BeforeEach
+    void setUp()
+    {
+        ReflectionTestUtils.setField( enrollmentService, "currentUserService", currentUserService );
```

Comment:
I think this is bad practice and was also the source of flaky tests in the past 😬 Note how you do not clean up your mutations. In integration (h2 or not) tests this will influence other tests with us not knowing why when it fails. 

If I understand correctly you are trying to test that given some `InstanceParam` we only fetch the necessary information.

If so can we write an integration test that shows that?

If not, then why not use `@InjectMocks` or call the constructor of an EnrollmentService where you pass in the mocked dependencies. If `@InjectMocks` does not work on an abstract class you can create a private subclass of it in this test class and test through that. Calling its constructor with the mocks.